File: common/src/main/java/com/vladte/devhack/common/service/domain/personalized/NoteService.java
package com.vladte.devhack.common.service.domain.personalized;

import com.vladte.devhack.common.service.domain.CrudService;
import com.vladte.devhack.entities.InterviewQuestion;
import com.vladte.devhack.entities.Note;
import com.vladte.devhack.entities.User;

import java.util.List;
import java.util.UUID;

/**
 * Service interface for Note entity operations.
 */
public interface NoteService extends CrudService<Note, UUID> {
    /**
     * Find notes by user.
     *
     * @param user the user
     * @return a list of notes by the user
     */
    List<Note> findNotesByUser(User user);

    /**
     * Find notes by linked question.
     *
     * @param question the question
     * @return a list of notes linked to the question
     */
    List<Note> findNotesByLinkedQuestion(InterviewQuestion question);

    /**
     * Find notes by user and linked question.
     *
     * @param user     the user
     * @param question the question
     * @return a list of notes by the user and linked to the question
     */
    List<Note> findNotesByUserAndLinkedQuestion(User user, InterviewQuestion question);
}

File: angular-ui/src/app/services/personalized/note.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {NoteDTO} from '../../models/personalized/note.model';
import {PersonalizedService} from '../personalized.service';

/**
 * Service for managing notes.
 * Extends PersonalizedService to inherit common CRUD operations for user-owned entities and follows OOP principles.
 */
@Injectable({
    providedIn: 'root'
})
export class NoteService extends PersonalizedService<NoteDTO> {
    protected baseUrl = '/api/notes';

    constructor(http: HttpClient) {
        super(http);
    }

    /**
     * Get current user's notes
     * @returns Observable array of user's notes
     */
    getMyNotes(): Observable<NoteDTO[]> {
        return this.http.get<NoteDTO[]>(`${this.baseUrl}/my-notes`);
    }

    /**
     * Get notes by question ID
     * @param questionId - Question ID to filter by
     * @returns Observable array of notes
     */
    getByQuestion(questionId: string): Observable<NoteDTO[]> {
        return this.http.get<NoteDTO[]>(`${this.baseUrl}/by-question/${questionId}`);
    }

    /**
     * Get notes by user ID
     * @param userId - User ID to filter by
     * @returns Observable array of notes
     */
    override getByUser(userId: string): Observable<NoteDTO[]> {
        return this.http.get<NoteDTO[]>(`${this.baseUrl}/by-user/${userId}`);
    }
}
